beacon_node/network/src/sync/sampling.rs

```diff
@@ -0,0 +1,486 @@
+        &mut self,
+        column_index: ColumnIndex,
+        result: Result<(), String>,
+        cx: &mut SyncNetworkContext<T>,
+    ) -> Result<Option<()>, SamplingError> {
+        // Select columns to sample
+        // Create individual request per column
+        // Progress requests
+        // If request fails retry or expand search
+        // If all good return
+        let Some(request) = self.column_requests.get_mut(&column_index) else {
+            warn!(
+                self.log,
+                "Received sampling response for unrequested column index"
+            );
+            return Ok(None);
+        };
+
+        match result {
+            Ok(_) => {
+                debug!(self.log, "Sample verification success"; "block_root" => %self.block_root, "column_index" => column_index);
+
+                // Valid, continue_sampling will maybe consider sampling succees
+                request.on_sampling_success()?;
+            }
+            Err(err) => {
+                debug!(self.log, "Sample verification failure"; "block_root" => %self.block_root, "column_index" => column_index, "reason" => ?err);
+
+                // TODO(das): Peer sent invalid data, penalize and try again from different peer
+                // TODO(das): Count individual failures
+                let peer_id = request.on_sampling_error()?;
+                cx.report_peer(
+                    peer_id,
+                    PeerAction::LowToleranceError,
+                    "invalid data column",
+                );
+            }
+        }
+
+        self.continue_sampling(cx)
+    }
+
+    fn continue_sampling(
+        &mut self,
+        cx: &mut SyncNetworkContext<T>,
+    ) -> Result<Option<()>, SamplingError> {
+        // First check if sampling is completed, by computing `required_successes`
+        let mut successes = 0;
+        let mut failures = 0;
+        let mut ongoings = 0;
+
+        for request in self.column_requests.values() {
+            if request.is_completed() {
+                successes += 1;
+            }
+            if request.is_failed() {
+                failures += 1;
+            }
+            if request.is_ongoing() {
+                ongoings += 1;
+            }
+        }
+
+        // If there are too many failures, consider the sampling failed
+        let Some(required_successes) = self.required_successes.get(failures) else {
+            return Err(SamplingError::TooManyFailures);
+        };
+
+        // If there are enough successes, consider the sampling complete
+        if successes >= *required_successes {
+            return Ok(Some(()));
+        }
+
+        let mut sent_requests = 0;
+
+        // First, attempt to progress sampling by requesting more columns, so that request failures
+        // are accounted for below.
+        for idx in 0..*required_successes {
+            // Re-request columns
+            let column_index = self.column_shuffle[idx];
```

Comment:
lint would probably complain about this, we should use `get` here to prevent index out of bounds.